compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala

```diff
@@ -399,45 +399,75 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
         } yield (argc, method)
       }
 
-      // Create a map: argCount -> methods (methods may appear multiple times)
-      val methodByArgCount =
-        methodArgCounts.groupBy(_._1).map(kv => kv._1 -> kv._2.map(_._2).toSet)
+      /** Like groupBy, but returns a sorted List instead of an unordered Map. */
+      def sortedGroupBy[A, K, O](xs: List[A])(grouper: A => K)(
+          sorter: ((K, List[A])) => O)(implicit ord: Ordering[O]): List[(K, List[A])] = {
+        xs.groupBy(grouper).toList.sortBy(sorter)
+      }
+
+      /* Create tuples: (argCount, methods).
+       * Methods may appear multiple times.
+       *
+       * The method lists preserve the order out of `methodArgCounts`, so if
+       * two such lists contain the same set of methods, they are equal.
+       *
+       * The resulting list is sorted by argCount. This is both for stability
+       * and because we then rely on the fact that the head is the minimum.
+       */
+      val methodByArgCount: List[(Int, List[Exported])] = {
+        sortedGroupBy(methodArgCounts)(_._1)(_._1) // sort by the Int
+          .map(kv => kv._1 -> kv._2.map(_._2)) // preserves the relative order of the Exported's
+      }
 
       // Create the formal args registry
-      val minArgc = methodByArgCount.keys.min
+      val minArgc = methodByArgCount.head._1 // it is sorted by argCount, so the head is the minimum
       val hasVarArg = varArgMeths.nonEmpty
       val needsRestParam = maxArgc != minArgc || hasVarArg
       val formalArgsRegistry = new FormalArgsRegistry(minArgc, needsRestParam)
 
       // List of formal parameters
       val (formalArgs, restParam) = formalArgsRegistry.genFormalArgs()
 
-      // Create tuples: (methods, argCounts). This will be the cases we generate
-      val caseDefinitions =
-        methodByArgCount.groupBy(_._2).map(kv => kv._1 -> kv._2.keySet)
+      /* Create tuples: (methods, argCounts). These will be the cases we generate.
+       *
+       * Within each element, the argCounts are sorted. (This sort order is
+       * inherited from the sort order of `methodByArgCount`.)
+       *
+       * For stability, the list as a whole is sorted by the minimum (head) of
+       * argCounts.
+       */
+      val caseDefinitions: List[(List[Exported], List[Int])] = {
+        sortedGroupBy(methodByArgCount)(_._2)(_._2.head._1) // sort by the minimum of the Ints
+          .map(kv => kv._1 -> kv._2.map(_._1)) // the Ints are still sorted from `methodByArgCount`
+      }
 
       // Verify stuff about caseDefinitions
       assert({
-        val argcs = caseDefinitions.values.flatten.toList
+        val argcs = caseDefinitions.map(_._2).flatten
         argcs == argcs.distinct &&
         argcs.forall(_ <= maxArgc)
       }, "every argc should appear only once and be lower than max")
 
+      def sameElementSet[A](xs: List[A], ys: List[A]): Boolean =
+        xs.size == ys.size && xs.forall(ys.contains(_))
```

Comment:
This is only correct if the lists contain distinct elements. I guess for all usages this is correct, but feels a bit brittle. Does it hurt performance significantly if we just do `xs.toSet == ys.toSet` (and maybe even inline at call site)?

Reply:
If the lists contain duplicates, we have bigger problems here. ;)

Reply:
I specialized it to testing for a `List[Exported]` against `varArgMeths`, with a comment reminding that they are distinct.